Adventure game rearranged into classes

Agata-Gizinska opened the pull request:
Adventure game refreshed version - created classes Player, Room (parent class), child classes: DarkRoom, MonsterRoom, Stairs, Prison, Basement; created game engine

Agata-Gizinska committed:
Adventure game refreshed version - created classes Player, Room (parent class), child classes: DarkRoom, MonsterRoom, Stairs, Prison, Basement; created game engine

czarnadzuma11 commented on adventure_game.py:
```diff
@@ -0,0 +1,473 @@
+        self.name = name
+        self.has_bottle = False
+        self.has_key = False
+        self._next_room = DarkRoom()  # create DarkRoom as initial room
```
Add initial_room=DarkRoom() paramter to __init__  and change this line to 
```self._next_room = initial_room```

Agata-Gizinska replied:
This solution throws an error:
Traceback (most recent call last):
  File "C:\Users\User\adventure_game.py", line 4, in <module>
    class Player:
  File "C:\Users\User\adventure_game.py", line 10, in Player
    def __init__(self, name='Nameless', initial_room=DarkRoom()):
NameError: name 'DarkRoom' is not defined

czarnadzuma11 replied:
Then modify the code for it to not throw an error, during code review you are usually presented with ideas for a suggested solution and are responsible to make it work. It's fine if you do not know how to address the issue but you should explain what you tried and request further instruction.

This happens because 
def __init__(..
gets evaluated when interpreter is reading the file and DarkRoom is not defined yet, interpreter needs it right now and we define it later down the file.

while 

self._next_room = DarkRoom()
gets evaluated later, when we actually start executing the function


One solution to such situation is to arrange the code in the order of dependency
class Room:
...
class DarkRoom: # below Room because it needs room to define itself
...
class Player: # below DarkRoom because it needs DarkRoom
...

but then we will run into another issue since Room class needs Player to be defined!
This in turn could be fixed by removing _player : Player type hint, which while may be nice is not critical and introduces a troublesome dependency.

The same is true for my previous idea of 
def init(self, name='Nameless', initial_room=DarkRoom()):
Rooms and players are not related in nature

What i think would be best to do is
1) remove room class name references in player class
2) remove player class name references in room class
3) make initial_room a required paramtere without default value. This way we will need to have all of the rooms and all of the players defined only when we will acutally need them, which currently is in main()

Agata-Gizinska replied:
done

czarnadzuma11 commented on adventure_game.py:
```diff
@@ -0,0 +1,473 @@
+
+    @staticmethod
+    def choice_room(player):
+        choice = int(input('Go left (1) or right (2)?'))
```
choice should keep asking the user for input until user provides an acceptable value. Create a function
```
def user_choice(prompt, choices):
  while True:
    choice = input(f'{prompt} {choices}')
    if choice in choices:
      return choice
```
and use it to ask for input, don't worry about types. Just use strings since you decide what those values are anyway

for example
```
choice = user_choice(prompt='Go left(1) or right(2)', ['1', '2']])
if choice == '1':
...
```

Agata-Gizinska replied:
ok

czarnadzuma11 commented on adventure_game.py:
```diff
@@ -0,0 +1,473 @@
+        self.first_choice(player)
+
+    def first_choice(self, player):
+        first_choice_ = str(input('Use stairs (yes/no)?'))
```
Give it a different name instead of suffixing with _, applies to the whole review

Agata-Gizinska replied:
ok

Agata-Gizinska committed:
Fix: changed naming instead of suffixing with _; added user_choice function which replaced all input calls in code

Agata-Gizinska committed:
Slow printing and fixes

feat:
- implemented slow_print function in order to slow down printing messages for user’s convenience and to force user to mark when the game can continue the story;
- enabled giving a custom name to the player;
fix:
- removed Room class references in Player class;
- removed Room class references in Player class;
- set initial room to be prompted from main function;
docs:
- implemented docstring for classes, parent classes’ methods, functions;
- adjusted to PEP 8

Agata-Gizinska committed:
Slow printing and fixes

feat:
- implemented slow_print function in order to slow down printing messages for user’s convenience and to force user to mark when the game can continue the story;
- enabled giving a custom name to the player;
fix:
- removed Room class references in Player class;
- removed Room class references in Player class;
- set initial room to be prompted from main function;
docs:
- implemented docstring for classes, parent classes’ methods, functions;
- adjusted to PEP 8

Agata-Gizinska committed:
Merge branch 'master' into devwork
